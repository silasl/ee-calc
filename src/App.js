import React, { Component } from 'react';
import styled from 'styled-components';
import { eval as MathEval } from 'mathjs';

import { GlobalStyle } from './global-styles';

import Brand from './components/brand';
import Screen from './components/screen';
import Button from './components/button';

const Wrapper = styled.div`
  width: 100%;
  max-width: 400px;
  background: #1795d4;
  padding: 10px;
`;

const Pad = styled.div`
  display: flex;
`;

const Numbers = styled.div`
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  margin-right: 20px;
`;

export const Number = styled(Button)`
  width: calc(33% - 5px);
`;

const Operators = styled.div`
  display: flex;
  flex-direction: column;
`;

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {input: ''};
  }

  clearInputValue = () => this.setState({ input: '' });
  setInputValue = (val) => this.setState({ input: `${this.state.input}${val}` });
  calculate = () => this.setState({ input: MathEval(this.state.input).toString() });

  render() {
    return (
      <Wrapper>
        <GlobalStyle />
        <Screen value={this.state.input} />
        <Brand>
          <Button value="C" handleClick={this.clearInputValue} />
        </Brand>
        <Pad>
          <Numbers>
            {['1','2','3','4','5','6','7','8','9', '0', '.'].map(val =>
              <Number value={val} handleClick={this.setInputValue} key={val} />
            )}
            <Number value={'='} handleClick={this.calculate} />
          </Numbers>
          <Operators>
              <Button value="/" label="รท" handleClick={this.setInputValue} />
              <Button value="*" label="x" handleClick={this.setInputValue} />
              <Button value="-" handleClick={this.setInputValue} />
              <Button value= "+" handleClick={this.setInputValue} />
          </Operators>
        </Pad>
      </Wrapper>
    );
  }
}

export default App;
