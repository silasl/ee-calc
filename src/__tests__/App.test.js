import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';

import Button from '../components/button';
import App, { Number } from '../App';

describe(App, () => {
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(<App />, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    it('should render correctly', () => {
        expect(shallow(<App />)).toMatchSnapshot();
    });

    it('should mount with the correct initial state', () => {
        const comp = shallow(<App />);
        expect(comp.state()).toEqual({ input: '' });
    });

    describe('State', () => {
        let comp;

        beforeEach(() => {
            comp = shallow(<App />);
        });

        it('should update the state as expected when a Number is clicked', () => {
            const btnClick = comp.find(Number).first().props().handleClick;
            btnClick('1');
            expect(comp.state()).toEqual({ input: '1' });
        });

        it('should concatenate the state input when Number is clicked', () => {
            comp.setState({ input: '1' });
            const btnClick = comp.find(Number).first().props().handleClick;
            btnClick('+');
            expect(comp.state()).toEqual({ input: '1+' });
        });

        it('should clear the state input when the clear button is clicked', () => {
            comp.setState({ input: 'foo' });
            comp.find(Button).first().props().handleClick();
            expect(comp.state()).toEqual({ input: '' });
        });

        it('should eval the input value and update the state when equals is clicked', () => {
            comp.setState({ input: '2+2' });
            comp.find(Number).last().props().handleClick();
            expect(comp.state()).toEqual({ input: '4' });
        });
    });
});