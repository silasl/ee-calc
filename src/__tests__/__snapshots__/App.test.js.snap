// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<styled.div>
  <GlobalStyleComponent />
  <Screen
    value=""
  />
  <Brand>
    <Button
      handleClick={[Function]}
      value="C"
    />
  </Brand>
  <styled.div>
    <styled.div>
      <Styled(Button)
        handleClick={[Function]}
        key="1"
        value="1"
      />
      <Styled(Button)
        handleClick={[Function]}
        key="2"
        value="2"
      />
      <Styled(Button)
        handleClick={[Function]}
        key="3"
        value="3"
      />
      <Styled(Button)
        handleClick={[Function]}
        key="4"
        value="4"
      />
      <Styled(Button)
        handleClick={[Function]}
        key="5"
        value="5"
      />
      <Styled(Button)
        handleClick={[Function]}
        key="6"
        value="6"
      />
      <Styled(Button)
        handleClick={[Function]}
        key="7"
        value="7"
      />
      <Styled(Button)
        handleClick={[Function]}
        key="8"
        value="8"
      />
      <Styled(Button)
        handleClick={[Function]}
        key="9"
        value="9"
      />
      <Styled(Button)
        handleClick={[Function]}
        key="0"
        value="0"
      />
      <Styled(Button)
        handleClick={[Function]}
        key="."
        value="."
      />
      <Styled(Button)
        handleClick={[Function]}
        value="="
      />
    </styled.div>
    <styled.div>
      <Button
        handleClick={[Function]}
        label="รท"
        value="/"
      />
      <Button
        handleClick={[Function]}
        label="x"
        value="*"
      />
      <Button
        handleClick={[Function]}
        value="-"
      />
      <Button
        handleClick={[Function]}
        value="+"
      />
    </styled.div>
  </styled.div>
</styled.div>
`;
